### ISSUE TITLE
M2T Usage

### ISSUE THREAD

**Issue by LeeYufang** [viatra/viatra-cps-benchmark#16]
Hi,
Sorry to disturb you again.
I have leart cps benchmark m2t for a long time and I have implemented m2t for another sample model with your means in cps benchmark. But It seems that your means is a little difficult.
Do you have a tutorial  or more related materials for m2t?

**Comment by abelhegedus** [comment]
We don't have a specific M2T tutorial, but if you would like to see a more complex example of a (partially) incremental M2T implementation, you could dive into the [EMDW-MC](https://github.com/IncQueryLabs/EMDW-MC) code.
I suggest you read the wiki and then start exploring the code related to the M2T part:
* [Code generation plugin](https://github.com/IncQueryLabs/EMDW-MC/tree/master/plugins/com.incquerylabs.emdw.cpp.codegeneration)
  * [Entry point](https://github.com/IncQueryLabs/EMDW-MC/blob/master/plugins/com.incquerylabs.emdw.cpp.codegeneration/src/com/incquerylabs/emdw/cpp/codegeneration/CPPCodeGeneration.xtend)
* [Tests](https://github.com/IncQueryLabs/EMDW-MC/tree/master/tests/com.incquerylabs.emdw.cpp.codegeneration.test)

**Comment by LeeYufang** [comment]
I have read the wiki and codes of  EMDW-MC M2T part.  For my understanding, end-user right clicks the menu "Generate CPP code for selected UML model" in model explore, the execution will go into "com.incquerylabs.emdw.cpp.ui.UmlHandler" , the M2M and M2T transformation will be  executed. Am I right?  If so , why the "toolchain.startChangeMonitor" task is put after the "toolchain.executeDeltaCodeAndFileGeneration" ?  Sorry, I am not completely understanding your work flow.  
        And if M2M and M2T both can be implemented by the incremental way with explicit traceability, can they use the same implementation process?
       Thanks and hope for your reply.

**Comment by abelhegedus** [comment]
> I have read the wiki and codes of EMDW-MC M2T part. For my understanding, end-user right clicks the menu "Generate CPP code for selected UML model" in model explore, the execution will go into [com.incquerylabs.emdw.cpp.ui.UmlHandler](https://github.com/IncQueryLabs/EMDW-MC/blob/master/plugins/com.incquerylabs.emdw.cpp.ui/src/com/incquerylabs/emdw/cpp/ui/UmlHandler.xtend) , the M2M and M2T transformation will be executed. Am I right?

Yes, that is correct. The generation itself can be observed in the [GeneratorJob](https://github.com/IncQueryLabs/EMDW-MC/blob/master/plugins/com.incquerylabs.emdw.cpp.ui/src/com/incquerylabs/emdw/cpp/ui/GeneratorJob.xtend#L137)

>  If so , why the "toolchain.startChangeMonitor" task is put after the "toolchain.executeDeltaCodeAndFileGeneration" ? Sorry, I am not completely understanding your work flow.

The same GeneratorJob is used for transforming the complete model and to incrementally update the generated code later. The first time we will go through the complete code generation, then start the change monitor to identify components that change later and use this information to run the code generation on the delta only.

> And if M2M and M2T both can be implemented by the incremental way with explicit traceability, can they use the same implementation process?
Thanks and hope for your reply.

The explicit traceability approach can be used if you can use a proper EMF model for traceability, while in an M2T usage, it is usually hard to create traceability to text artifacts. It is probably possible to use the same "process" if you create some kind of detailed traceability information, but for the EMDW-MC toolchain, we opted to create a "dirty incremental" M2T step, where the change monitor examines model modifications on the source model to identify affected high-level elements (Components) and the code generation can be re-executed on those elements (instead of the complete model).
